### PR TITLE
Opt/connected components

### PR DESCRIPTION
**Pull request by jpfairbanks**
I added some keyword arguments in order to allow for even more manual memory management.
reduces the total allocation on `SNAP/roadNet-CA` 128GB to 47MB. As they say: "If you want memory managed right, you have to do it yourself"

```
Working on SNAP/roadNet-CA
Reading MTX of SNAP/roadNet-CA
Symmetrizing 
elapsed time: 0.00484772 seconds
Finding Components
 17.076791 seconds (43.56 k allocations: 47.113 MB, 0.01% gc time)
 42.093054 seconds (81.34 k allocations: 127.992 GB, 6.30% gc time)
  0.313385 seconds (18 allocations: 15.540 MB, 0.18% gc time)
```

### CODE REVIEW

**Comment by jpfairbanks** [comment]
Lots of tiny components previously cause my version to take quadratic time.
Instead of building the tree and then using the tree to identify the components, I wrote a ComponentBFSVisitorVector to just compute the components without building the trees.
It is about the same for graphs with 1 component but much faster on graphs with many components

```
Reading MTX of SNAP/amazon0302
Symmetrizing 
the matrix is not symmetric using A+A'
elapsed time: 0.051966621 seconds
Finding Components
  0.069624 seconds (6 allocations: 4.000 MB)
  0.071326 seconds (43 allocations: 8.251 MB)
  0.070003 seconds (21 allocations: 5.000 MB)
length(components) = 1

Working on SNAP/roadNet-CA
Reading MTX of SNAP/roadNet-CA
Symmetrizing 
elapsed time: 0.004767097 seconds
Finding Components
  0.330907 seconds (26.14 k allocations: 31.143 MB, 0.69% gc time)
 47.077985 seconds (78.72 k allocations: 127.992 GB, 5.58% gc time)
  0.333240 seconds (18 allocations: 15.540 MB, 0.18% gc time)
length(components) = 8713
```

File: test/traversals/bfs.jl [hunk 1]
```diff
@@ -1,5 +1,8 @@
 z = bfs_tree(g5, 1)
+visitor = LightGraphs.TreeBFSVisitorVector(zeros(Int,nv(g5)))
+t = LightGraphs.bfs_tree(visitor, g5, 1)
```
Comment (sbromberger) [inline test/traversals/bfs.jl]: Why do we need to specify the module here?

File: test/connectivity.jl [hunk 2]
```diff
@@ -10,6 +10,12 @@ add_edge!(g,10,9)
 @test is_connected(HouseGraph())
 
 cc = connected_components(g)
+visitor = LightGraphs.TreeBFSVisitorVector(zeros(Int, nv(g)))
+label = zeros(Int, nv(g))
+label = LightGraphs.connected_components!(label, g)
```
Comment (sbromberger) [inline test/connectivity.jl]: Why do we need to specify the module here?

File: bench/fill.jl [hunk 3]
```diff
@@ -0,0 +1,41 @@
+using Benchmarks
```
Comment (sbromberger) [inline bench/fill.jl]: Are these necessary for LightGraphs? They look like generic benchmark tests.

File: test/connectivity.jl (hunk 2 above)
Comment (jpfairbanks) [inline test/connectivity.jl]: I didn't add it as an export. Would you like to export it?

Comment (sbromberger) [inline test/connectivity.jl]: Oh, I didn't see the exclamation point. Got it. If it's called by `connected_components` then we don't need to. What do you prefer for an API?

Comment (jpfairbanks) [inline test/connectivity.jl]: `connected_components!` produces a vector of labels where the label is the root of the BFS tree that found the vertex. This API is better for parallel access than the Vector{Vector{Int}} that the `connected_components` uses. We could wrap the labels version to support the api of the standard version. Then experts could use this directly.

**Comment by sbromberger** [comment]
The travis failures are on 0.3. Is there any advantage to merging this into sparsemx instead of keeping it separate? The reason is we're going to be dropping 0.3 support if we adopt sparsemx.

**Comment by sbromberger** [comment]
heads up to @ZacCranko - we're looking at making a (breaking) change to connected components here.

**Comment by sbromberger** [comment]
This is cool. I could see something like

```
x = components(label)
f = Dict{Int, Graph}
for k in keys()
    f[k] = g[x[k]]
end
```
to give subgraphs of the components.

**Comment by jpfairbanks** [comment]
So I replaced the slow version with the following cycle:
```
setup for fast version
call fast version
convert output to old api
```

The only breaking change is that you now get singletons.
If all vertices in your graph have nonzero degree, then there will be no change to you.
If your code does not expect singletons, you will have to change it either by filtering out the singletons with a `filter(x->length(x)>1, connected_components(g))`, or by changing your code to do the right thing for singletons.

While I don't like the idea of providing a different data structure in the low level and high level api, I think this is a fair compromise between ease of programming and performance. If anyone wants to use the faster data structure (flat array of labels) then they can do that.

This should be a nice performance boost for all applications. And a huge speedup for sparse graphs with many connected components.

**Comment by sbromberger** [comment]
Yes - this is much much faster on sparse graphs. Thank you for all your work on this. Can't wait to merge it :)

**Comment by jpfairbanks** [comment]
Checks pass! Ship it! :8ball:

**Comment by sbromberger** [comment]
Love it. Love it. Love it. Thanks!!

**Comment by ZacCranko** [comment]
Doesn't sound like it'll break any of our stuff. Thanks for the heads up though.

**Pull request merged**